ouroboros-consensus-shelley/src/Ouroboros/Consensus/Shelley/ShelleyHFC.hs

```diff
@@ -0,0 +1,311 @@
+{-# LANGUAGE DataKinds             #-}
+{-# LANGUAGE DeriveAnyClass        #-}
+{-# LANGUAGE DeriveGeneric         #-}
+{-# LANGUAGE FlexibleContexts      #-}
+{-# LANGUAGE FlexibleInstances     #-}
+{-# LANGUAGE MultiParamTypeClasses #-}
+{-# LANGUAGE RecordWildCards       #-}
+{-# LANGUAGE ScopedTypeVariables   #-}
+{-# LANGUAGE TypeApplications      #-}
+{-# LANGUAGE TypeFamilies          #-}
+{-# LANGUAGE TypeOperators         #-}
+{-# LANGUAGE UndecidableInstances  #-}
+{-# OPTIONS_GHC -Wno-orphans #-}
+module Ouroboros.Consensus.Shelley.ShelleyHFC (
+    ShelleyBlockHFC
+  , ShelleyPartialLedgerConfig (..)
+  , ProtocolShelley
+  , forecastAcrossShelley
+  , translateChainDepStateAcrossShelley
+  , translateLedgerViewAcrossShelley
+  ) where
+
+import           Control.Monad (guard)
+import           Control.Monad.Except (Except, throwError)
+import qualified Data.Map.Strict as Map
+import           Data.Maybe
+import           Data.SOP.Strict
+import           Data.Void (Void)
+import           Data.Word
+import           GHC.Generics (Generic)
+import           NoThunks.Class (NoThunks)
+
+import           Ouroboros.Consensus.Block
+import           Ouroboros.Consensus.Forecast
+import           Ouroboros.Consensus.HardFork.History (Bound (boundSlot))
+import           Ouroboros.Consensus.HardFork.Simple
+import           Ouroboros.Consensus.Ledger.Abstract
+import           Ouroboros.Consensus.TypeFamilyWrappers
+
+import           Ouroboros.Consensus.Config
```

Comment:
This group, the group above it, and the group below it should all be merged; no need for the lines between them.